moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
+      //
+      try {
+        Method getConstructorId = ObjectStreamClass.class.getDeclaredMethod(
+            "getConstructorId", Class.class);
+        getConstructorId.setAccessible(true);
+        final int constructorId = (Integer) getConstructorId.invoke(null, Object.class);
+        final Method newInstance = ObjectStreamClass.class.getDeclaredMethod("newInstance",
+            Class.class, int.class);
+        newInstance.setAccessible(true);
+        return new ClassFactory<T>() {
+          @SuppressWarnings("unchecked")
+          @Override
+          public T newInstance() throws InvocationTargetException, IllegalAccessException {
+            return (T) newInstance.invoke(null, rawType, constructorId);
+          }
+          @Override public String toString() {
+            return rawType.getName();
+          }
+        };
+      } catch (IllegalAccessException e) {
+        throw new AssertionError();
+      } catch (InvocationTargetException e) {
+        throw Util.rethrowCause(e);
+      } catch (NoSuchMethodException ignored) {
+        // Not the expected version of Dalvik/libcore!
+      }
+
     }
 
-    // Try (pre-Gingerbread) Dalvik/libcore's ObjectInputStream mechanism.
-    // public class ObjectInputStream {
-    //   private static native Object newInstance(
-    //     Class<?> instantiationClass, Class<?> constructorClass);
-    // }
-    try {
-      final Method newInstance = ObjectInputStream.class.getDeclaredMethod(
-          "newInstance", Class.class, Class.class);
-      newInstance.setAccessible(true);
-      return new ClassFactory<T>() {
-        @SuppressWarnings("unchecked")
-        @Override public T newInstance() throws InvocationTargetException, IllegalAccessException {
-          return (T) newInstance.invoke(null, rawType, Object.class);
-        }
-        @Override public String toString() {
-          return rawType.getName();
-        }
-      };
-    } catch (Exception ignored) {
+    if (isBelowAndroidApi(9)) {
+      // Try (pre-Gingerbread) Dalvik/libcore's ObjectInputStream mechanism.
+      // public class ObjectInputStream {
+      //   private static native Object newInstance(
+      //     Class<?> instantiationClass, Class<?> constructorClass);
+      // }
+      try {
+        final Method newInstance = ObjectInputStream.class.getDeclaredMethod(
+            "newInstance", Class.class, Class.class);
+        newInstance.setAccessible(true);
+        return new ClassFactory<T>() {
+          @SuppressWarnings("unchecked")
+          @Override
+          public T newInstance() throws InvocationTargetException, IllegalAccessException {
+            return (T) newInstance.invoke(null, rawType, Object.class);
+          }
+          @Override public String toString() {
+            return rawType.getName();
+          }
+        };
+      } catch (Exception ignored) {
+      }
     }
 
     throw new IllegalArgumentException("cannot construct instances of " + rawType.getName());
```

Comment:
Since we have no solution for API 29+, maybe we should throw a more instructive exception in that case (pointing them to the workaround, i.e. "you are on API 29+, you need an empty constructor")?

Reply:
Good call - 40b2ed3